### PR TITLE
[fix](Nereids) some expression not cast in InPredicate

### PR DESCRIPTION
**Pull request by morrySnow**
1. should use castIfNotSameType in InPredicate and CaseWhen
2. castIfNotSameType should cast all type if two side is not exactly same

## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by morrySnow** [comment]
run buildall

**Comment by morrySnow** [comment]
run buildall

**Comment by morrySnow** [comment]
run buildall

**Comment by morrySnow** [comment]
run buildall

**Comment by morrySnow** [comment]
run buildall

**Comment by morrySnow** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/SimplifyInPredicate.java [hunk 1]
```diff
@@ -46,12 +47,25 @@ public Expression visitInPredicate(InPredicate expr, ExpressionRewriteContext co
...
-                        List<Expression> children = Lists.newArrayList();
+                        ImmutableList.Builder<Expression> children = ImmutableList.builder();
                         children.add(cast.child());
-                        literals.stream().forEach(
-                                l -> children.add(convertToDateV2Literal((DateTimeV2Literal) l)));
-                        return expr.withChildren(children);
+                        literals.forEach(l -> children.add(convertToDateV2Literal((DateTimeV2Literal) l)));
+                        return expr.withChildren(children.build());
                     }
+                } else if (cast.child().getDataType().isDateTimeV2Type()
```
Comment (englefly) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/SimplifyInPredicate.java:55]: cast(A as xxx) in (opt1, opt2, ...)
if A is DatetimeV2 type, and opt1 is DateTimeV2Literal, why there is a cast?

**Review by starocean999 — APPROVED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/SimplifyInPredicate.java (hunk 1 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/SimplifyInPredicate.java:55]: u could see example from UT `SimplifyInPredicateTest.java`

**Review by englefly — APPROVED** [review]

**Pull request merged**